Clarify error message when url file doesn't exists

bogdasar1985 opened the pull request:
Fix of https://github.com/newsboat/newsboat/issues/1919

der-lyse reviewed (changes requested):
In my opinion there's no need to change all the PO files, that's part of the release process. But if so, all the touched and not translated messages should be marked `fuzzy`, otherwise translators will not find them.

der-lyse reviewed (commented):
Thank you very much! This looks quite good. Taking a deeper look now at the `read_text_file(…)` function I'm wondering whether this new hint should only be appended in case of `CantOpen`, but not `LineError`. So the hint might be better appended to the message `"Failed to open file: %s"` instead.

But I'm no Rust programmer at all and don't consider myself a C++ developer either, so maybe I'm overlooking something. It could well be that the `LineError` case would also benefit from this new hint. The Rust docs of the `LineError` just say:

> One of the lines triggered an io::Error (most likely invalid UTF-8, but see `reason` for details).

In case of invalid UTF-8 the hint on file existence and read permissions doesn't help. Not sure, what else might cause a line error, though. @Minoru or @dennisschagt, any advice on this here?

Minoru commented:
> Taking a deeper look now at the read_text_file(…) function I'm wondering whether this new hint should only be appended in case of CantOpen, but not LineError.

It goes even higher than this: `FileUrlReader` doesn't know in which context it's used, and thus can't display an appropriate message. When it's used as part of a normal startup, sure, it should say "please create this file and add URLs there". When it's used as part of OPML import, the message should say something about creating ~/.newsboat, so Newsboat can put the file there.

I already explained this in https://github.com/newsboat/newsboat/issues/1919#issuecomment-1007971673. I understand that the fix I'm suggesting is more involved, but I think this PR doesn't fix the bug at all: instead of no message we now have a misleading message.

@bogdasar1985, would you like to work on a solution that makes `FileUrlReader` return an error code instead of an error message, so that the caller can decide what message to display?

bogdasar1985 commented:
@Minoru 

> @bogdasar1985, would you like to work on a solution that makes FileUrlReader return an error code instead of an error message, so that the caller can decide what message to display?

I gonna try, will see what I able to do.

bogdasar1985 commented:
@Minoru

1) I changed `FileUrlReader::reload()` method signature. Now it returns `ReadTextFileError` structure, which contains enum with error kind and message.

2) Make error message, which formed in `int Controller::import_opml` more understandable.

3) I delete `return EXIT_FAILURE;` from here https://github.com/bogdasar1985/newsboat/blob/6c4734875df9c2845dcd9ad7ed0462fe7169a165/src/controller.cpp#L365
In that case program fail a little later, managing to display a message that the file was not found, like here https://newsboat.org/releases/2.26/docs/newsboat.html#_first_steps

bogdasar1985 commented:
@Minoru and most importantly
at the moment we got next behaviour:
```
bogdasar@bogdasar-System-Product-Name:~/Documents/Programming/C++_Programming/newsboat$ ./newsboat 
Запускаю Newsboat r2.26-22-g6c473...
Загружаю настройки...сделано.
Открываю кэш...сделано.
Загружаю ссылки из /home/bogdasar/.newsboat/urls...Не удалось открыть файл: Нет такого файла или каталога (os error 2)

Ошибка: не найдено ни одной ссылки. Пожалуйста, занесите в файл /home/bogdasar/.newsboat/urls ссылки на RSS ленты или импортируйте файл OPML.


bogdasar@bogdasar-System-Product-Name:~/Documents/Programming/C++_Programming/newsboat$ ./newsboat --import-from-opml rss-subscriptions.opml 
An error occured while importing OPML to url file /home/bogdasar/.newsboat/urls : Не удалось открыть файл: Нет такого файла или каталога (os error 2)

```

Minoru commented:
Sorry, I don't have time to review this right now. I'll do it on Sunday. The new behaviour looks like exactly what I wanted though, that's great!

The build errors are most likely due to new `#include`s. Please run `make deps` and commit the result, or copy the diff from [this CI job](https://github.com/newsboat/newsboat/pull/1925/checks?check_run_id=4801829071) You can also run `make distclean` and re-run `make` to get the same errors locally.

bogdasar1985 commented:
@Minoru 
`make deps` output that
```
bogdasar@bogdasar-System-Product-Name:~/Documents/Programming/C++_Programming/newsboat$ make deps
Checking for package sqlite3... found
Checking for package libcurl... found
Checking for package libxml-2.0... found
Checking for package stfl... found
Checking for command asciidoctor... found
Checking for command cargo... /usr/bin/asciidoctor
found
/usr/bin/cargo
Checking for package json... not found
Checking for package json-c... found
Checking for package ncursesw... found
/usr/bin/curl-config: 1: krb5-config: not found
/usr/bin/curl-config: 1: krb5-config: not found
make: *** Нет правила для сборки цели «deps».  Останов.
```
but only if before that you run `make distclean`, if not, you got that:
```
bogdasar@bogdasar-System-Product-Name:~/Documents/Programming/C++_Programming/newsboat$ make deps
make: *** Нет правила для сборки цели «deps».  Останов.
```

Minoru commented:
Right, sorry, I meant depslist! It should have modified some files - please commit them to fix CI.

bogdasar1985 commented:
@Minoru ok, test broken, because I changed signature. Well, let's fix it.

bogdasar1985 commented:
@Minoru so, I commented the sections where "contains filename" was tested, because now return value don't contain filename. First test, I changed to another, second one, I deleted at all.

Minoru commented:
Hmm, the behaviour now differs from what you've shown in https://github.com/newsboat/newsboat/pull/1925/#issuecomment-1012046415:

```
$ export HOME=$(mktemp -d)
$ ./newsboat
Starting Newsboat r2.26-30-g4350...
Loading configuration...done.
Opening cache...done.
Loading URLs from /tmp/tmp.YDeeOmvBNe/.newsboat/urls...Failed to open file: No such file or directory (os error 2)

Error: no URLs configured. Please fill the file /tmp/tmp.YDeeOmvBNe/.newsboat/urls with RSS feed URLs or import an OPML file.

$ export HOME=$(mktemp -d)
$ newsboat --import-from-opml rss-subscriptions.opml
Import of rss-subscriptions.opml finished.
```

Newsboat now creates ~/.newsboat/urls, but it should have displayed an error.

Looks like the new error message is commented out entirely? Please get this back into a state described in your comment.

Also, please don't bother updating the PO files. As I said eralier, it will get done automatically before the next release.

Minoru commented:
Actually, I am not sure if the automation for PO files will work after your changes, because the POT file wasn't updated. Please undo your changes to the po directory.

bogdasar1985 commented:
@Minoru 

> Newsboat now creates ~/.newsboat/urls, but it should have displayed an error.

No. I tried in my PC, all is ok. And everything work, as I described.

Maybe, it's because we run a different versions. I run that:
```
bogdasar@bogdasar-Aspire-V3-571G:/home/bogdasar/Документы/Programming/C++_Programming/newsboat$ ./newsboat -v
Newsboat r2.26-30-g4350 - https://newsboat.org/
Copyright (C) 2006-2015 Andreas Krennmair
Copyright (C) 2015-2021 Alexander Batischev
Copyright (C) 2006-2017 Newsbeuter contributors
Copyright (C) 2017-2021 Newsboat contributors

Newsboat является свободным программным обеспечением под лицензией MIT. (Для просмотра полного текста наберите `./newsboat -vv').
Программа включает в себя:
- библиотеку «JSON for Modern C++» под лицензией MIT: https://github.com/nlohmann/json
- библиотеку «optional-lite» под лицензией Boost Software License: https://github.com/martinmoene/optional-lite
- библиотеку «expected-lite» под лицензией Boost Software License: https://github.com/martinmoene/expected-lite

Newsboat r2.26-30-g4350
System: Linux 5.4.0-91-generic (x86_64)
Compiler: g++ 9.3.0
ncurses: ncurses 6.2.20200212 (compiled with 6.2)
libcurl: libcurl/7.68.0 GnuTLS/3.6.13 zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3 (compiled with 7.68.0)
SQLite: 3.31.1 (compiled with 3.31.1)
libxml2: compiled with 2.9.10



bogdasar@bogdasar-Aspire-V3-571G:/home/bogdasar/Документы/Programming/C++_Programming/newsboat$ git remote -v
origin	git@github.com:bogdasar1985/newsboat.git (fetch)
origin	git@github.com:bogdasar1985/newsboat.git (push)



bogdasar@bogdasar-Aspire-V3-571G:/home/bogdasar/Документы/Programming/C++_Programming/newsboat$ git log | head -n 6
commit 83b19ae72ef88b4524f5c8c3a41ff49450c4d056
Author: bogdasar <bogdasar1985@gmail.com>
Date:   Mon Jan 17 01:05:22 2022 +0500

    Undo po files changes

```

bogdasar1985 commented:
@Minoru look, first time you run `$ ./newsboat`, but second time you run `$ newsboat`. It's different things! Maybe it's a reason?

Minoru commented:
You're totally right! Yes, the behaviour does match your comment. Let me do a review then…

Minoru commented on include/fileurlreader.h:
```diff
@@ -4,6 +4,7 @@
 #include <string>
 
 #include "urlreader.h"
+#include "utils.h"
```
Please remove this line, as utils.h is already included by urlreader.h.

Minoru commented on include/fileurlreader.h:
```diff
@@ -14,7 +15,7 @@ class FileUrlReader : public UrlReader {
 	///
 	/// \return A non-value on success, an error message otherwise.
-	nonstd::optional<std::string> reload() override;
+	nonstd::optional<utils::ReadTextFileError> reload() override;
```
Please update the docstring.

Minoru commented on src/controller.cpp:
```diff
@@ -723,7 +723,10 @@ int Controller::import_opml(const std::string& opmlFile,
 	if (error_message.has_value()) {
-		std::cout << error_message.value() << std::endl;
+		std::cout << strprintf::fmt(
+				_("An error occured while importing OPML to url file %s : %s"),
```
```suggestion
				_("Error importing OPML to urls file %s: %s"),
```

This is shorter, giving more space for paths and the clarification; it also uses terminology and punctuation the rest of Newsboat does ;)

Minoru commented on src/controller.cpp:
```diff
@@ -360,8 +360,8 @@ int Controller::run(const CliArgsParser& args)
-		std::cout << error_message.value() << std::endl << std::endl;
-		return EXIT_FAILURE;
+		std::cout << error_message.value().message << std::endl << std::endl;
+		//return EXIT_FAILURE;
```
Don't comment out code, just remove it ;) We have Git, so we can restore anything at any time.

Minoru commented on src/fileurlreader.cpp:
```diff
@@ -17,17 +18,21 @@ std::string FileUrlReader::get_source()
 	auto result = utils::read_text_file(filename);
 	if (!result) {
-		return strprintf::fmt(_("Error: failed to read URLs from file \"%s\": %s"),
+		/**
+		return strprintf::fmt(
+				_("Error: failed to read URLs from file \"%s\": %s. Please make sure the file exists and you have read permissions."),
 				filename,
 				result.error().message);
+		*/
```
Please remove the comment, we have the old code in Git history.

Minoru commented on test/fileurlreader.cpp:
```diff
@@ -143,11 +149,12 @@ TEST_CASE("URL reader returns error message if file cannot be opened",
-
+			/**
 			SECTION("the error message contains the filename") {
-				INFO("error_message: " + error_message.value());
-				REQUIRE(error_message.value().find(urlsFile.get_path()) != std::string::npos);
+				INFO("error_message: " + error_message.value().message);
+				REQUIRE(error_message.value().message.find(urlsFile.get_path()) != std::string::npos);
 			}
+			*/
```
I agree that this section no longer applies. Please remove it.

Minoru commented on test/fileurlreader.cpp:
```diff
@@ -102,21 +103,26 @@ TEST_CASE("Preserves URLs as-is", "[FileUrlReader][issue926]")
-
+		/**
 		SECTION("the error message contains the filename") {
-			INFO("error_message: " + error_message.value());
-			REQUIRE(error_message.value().find(urlsFile.get_path()) != std::string::npos);
+			INFO("error_message: " + error_message.value().message);
+			REQUIRE(error_message.value().message.find(urlsFile.get_path()) != std::string::npos);
+		}
+		*/
```
Please drop this comment; the new `SECTION` replaces it.

Minoru commented on test/fileurlreader.cpp:
```diff
@@ -102,21 +103,26 @@ TEST_CASE("Preserves URLs as-is", "[FileUrlReader][issue926]")
+		}
+		*/
+		SECTION("the error structure contains error message and kind of an error") {
+			INFO("error_message: " + error_message.value().message);
```
Please change this to a `REQUIRE` that checks that the message is non-empty; it's better than logging because it's automated.

Minoru reviewed (changes requested):
This is getting ready, thank you for your work! My comments are mostly about removing stuff that is no longer needed, should be easy to address :)

Minoru commented on src/controller.cpp:
```diff
@@ -724,7 +723,7 @@ int Controller::import_opml(const std::string& opmlFile,
 	if (error_message.has_value()) {
 		std::cout << strprintf::fmt(
-				_("An error occured while importing OPML to url file %s : %s"),
+				_("Error importing OPML to urls file %s : %s"),
```
```suggestion
				_("Error importing OPML to urls file %s: %s"),
```

Minoru commented on test/fileurlreader.cpp:
```diff
@@ -114,15 +114,9 @@ TEST_CASE("URL reader returns error structure if file cannot be opened",
 		SECTION("the error structure contains error message and kind of an error") {
 			INFO("error_message: " + error_message.value().message);
-			REQUIRE(error_message.value().kind == utils::ReadTextFileErrorKind::CantOpen);
+			REQUIRE(!error_message.value().message.empty());
```
```suggestion
			REQUIRE_FALSE(error_message.value().message.empty());
```

This catches more attention and is thus more idiomatic for this particular test framework.

Minoru reviewed (approved):
Thank you!

Minoru commented:
I believe @der-lyse's review comments were addressed, so I'll dismiss his request for changes now.
